jobs/integration/conftest.py

```diff
@@ -143,41 +143,49 @@ def pytest_addoption(parser):
-        request = self._request
-        whoami, _ = await self.run("juju", "whoami", "--format=yaml")
-        stdout, _ = await self.run("juju", "--version")
+    def _load(self):
+        whoami, _ = self.exec("juju", "whoami", "--format=yaml")
+        stdout, _ = self.exec("juju", "--version")
         ver_str = stdout.splitlines()[-1].split("-")[0]
         self.juju_version = tuple(map(int, ver_str.split(".")))
         self.juju_user = yaml.safe_load(whoami)["user"]
-        self.controller_name = request.config.getoption("--controller")
-        self.model_name = request.config.getoption("--model")
+        self.controller_name = self._config.getoption("--controller")
+        self.model_name = self._config.getoption("--model")
         self.model_name_full = f"{self.juju_user}/{self.model_name}"
         self.k8s_model_name = f"{self.model_name}-k8s"
         self.k8s_model_name_full = f"{self.model_name_full}-k8s"
-        self.series = request.config.getoption("--series")
-        self.cloud = request.config.getoption("--cloud")
+        self.series = self._config.getoption("--series")
         self.k8s_cloud = f"{self.k8s_model_name}-cloud"
         self.connection = f"{self.controller_name}:{self.model_name_full}"
         self.k8s_connection = f"{self.controller_name}:{self.k8s_model_name_full}"
-        self.is_series_upgrade = request.config.getoption("--is-series-upgrade")
+        self.is_series_upgrade = self._config.getoption("--is-series-upgrade")
         self.charm_channel = (
-            request.config.getoption("--charm-channel")  # use specified channel
+            self._config.getoption("--charm-channel")  # use specified channel
             or os.environ.get("CHARM_CHANNEL_UPGRADE_TO")  # fallback to upgrade env var
             or os.environ.get("JUJU_DEPLOY_CHANNEL")  # fallback to env var
             or "edge"  # default to edge
         )
-        self.snap_channel = request.config.getoption("--snap-channel")
-        self.vault_unseal_command = request.config.getoption("--vault-unseal-command")
-        self.juju_ssh_proxy = request.config.getoption("--juju-ssh-proxy")
-        self.use_existing_ceph_apps = request.config.getoption(
-            "--use-existing-ceph-apps"
+        self.snap_channel = self._config.getoption("--snap-channel")
+        self.vault_unseal_command = self._config.getoption("--vault-unseal-command")
+        self.juju_ssh_proxy = self._config.getoption("--juju-ssh-proxy")
+        self.use_existing_ceph_apps = self._config.getoption("--use-existing-ceph-apps")
+
+    @cached_property
+    def cloud(self):
+        if _it := self._config.getoption("--cloud"):
+            return _it
+        controller_data, _ = self.exec(
+            "juju", "show-controller", self.controller_name, "--format", "yaml"
         )
+        controller_infos = yaml.safe_load(controller_data)
+        controller_info, *_ = controller_infos.values()
+        return controller_info["details"]["cloud"]
```

Comment:
the cloud field can be loaded from the existing **required** controller name argument.